### PR TITLE
[SequenceFeatureExtractor] Rewrite padding logic from pure python to numpy

### LINKED ISSUE #13539: [SequenceFeatureExtraction] Move padding logic from pure python to numpy
Currently the padding for sequential feature extraction: https://github.com/huggingface/transformers/blob/9d60eebeb52ed3c266ab8e0cc6871ebeb08a5bc1/src/transformers/feature_extraction_sequence_utils.py#L38 is done in pure Python even though the two feature extractors that rely on sequential both always return either a single numpy (or PT or TF) tensor or a list of numpy tensors - see:

- https://github.com/huggingface/transformers/blob/9d60eebeb52ed3c266ab8e0cc6871ebeb08a5bc1/src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py#L205 for Wav2Vec2
- https://github.com/huggingface/transformers/blob/9d60eebeb52ed3c266ab8e0cc6871ebeb08a5bc1/src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py#L209 

More specifically: 

1) `feature_extractor(raw_audio, return_tensors=None, padding=True)` => returns a list of  numpy arrays
2) `feature_extractor(raw_audio, return_tensors="np", padding=True)` => returns a 2D numpy array (PT and TF return a 2D array in PT and TF correspondingly)

For some background the text tokenizers all have the option to return "pure" Python lists which was then implemented analogous for the speech feature extractors. However it was decided that the each individual sequence will always have be a numpy array. This is because a very common use case is to load audio files via `torchaudio` or `librosa` which return a list of numpy arrays and then run the feature extractor on it. Running `feature_extractor(list_of_numpy_arrays, padding=True)` should then not change the inner numpy arrays to pure python as this would be confusing for the user and also very be very slow. 

Therefore I think the current design is good as is and that we can move the padding logic in https://github.com/huggingface/transformers/blob/9d60eebeb52ed3c266ab8e0cc6871ebeb08a5bc1/src/transformers/feature_extraction_sequence_utils.py#L38 from "pure" Python to a numpified approach.. This would both improve speed and prevent instability issues like: https://github.com/huggingface/transformers/pull/13538 . This won't change anything in the user-facing API since at the moment the the "inner" output is always a numpy vector anyways. 

Do you agree on this @LysandreJik @sgugger @anton-l @patil-suraj  ?

@anton-l - are you maybe interested in giving this PR a try?

### PR DESCRIPTION
**Pull request by anton-l**
# What does this PR do?

Resolves #13539 

Since speech models universally use Numpy `float32` arrays as input features (standard way of representing waveforms), it was decided to rewrite `SequenceFeatureExtractor` from pure python lists (akin to traditional tokenizers) to numpy arrays. It will also help with solving some inconsistent normalization issues (#13538, #13585) due to `float->np.float32` conversions.

The feature extractor itself is still dtype-agnostic (can pad `np.float64` in the future if needed), while the model-specific feature extractors were updated to only work with `np.float32`

### CODE REVIEW

File: src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py [hunk 1]
```diff
@@ -99,15 +99,12 @@ def utterance_cmvn(
...
-        square_sums = (x[:input_length] ** 2).sum(axis=0)
-
+        # make sure we normalize float32 arrays
         if normalize_means:
+            mean = x[:input_length].mean(axis=0)
             x = np.subtract(x, mean)
         if normalize_vars:
-            var = square_sums / x[:input_length].shape[0] - mean ** 2
-            std = np.sqrt(np.maximum(var, 1e-10))
+            std = x[:input_length].std(axis=0)
```
Comment (anton-l) [inline src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:107]: Switched this logic to pure numpy to squeeze out a bit more precision when working with `np.float32` instead of casted `float->np.float64`

File: src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py [hunk 2]
```diff
@@ -206,10 +203,10 @@ def __call__(
             and (isinstance(raw_speech[0], np.ndarray) or isinstance(raw_speech[0], (tuple, list)))
         )
 
-        if is_batched and not isinstance(raw_speech[0], np.ndarray):
-            raw_speech = [np.asarray(speech) for speech in raw_speech]
+        if is_batched:
+            raw_speech = [np.asarray(speech, dtype=np.float32) for speech in raw_speech]
```
Comment (anton-l) [inline src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:207]: Forcing float32 here for consistency

File: src/transformers/feature_extraction_sequence_utils.py [hunk 3]
```diff
@@ -278,42 +280,47 @@ def _pad(
...
         if needs_to_be_padded:
             difference = max_length - len(required_input)
-            padding_vector = self.feature_size * [self.padding_value] if self.feature_size > 1 else self.padding_value
             if self.padding_side == "right":
                 if return_attention_mask:
-                    processed_features["attention_mask"] = [1] * len(required_input) + [0] * difference
-                processed_features[self.model_input_names[0]] = required_input + [
-                    padding_vector for _ in range(difference)
-                ]
+                    attention_mask = np.zeros(max_length, dtype=np.int32)
```
Comment (sgugger) [inline src/transformers/feature_extraction_sequence_utils.py:289]: Why int32? Usually we cast everything to int64 in data collators as some layers are not implemented in int32
And if it's to reduce memory footprint, let's go for a bool dtype.

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

**Review by patil-suraj — APPROVED** [review]
LGTM!

File: tests/test_modeling_speech_to_text.py [hunk 4]
```diff
@@ -724,7 +724,7 @@ def map_to_array(batch):
             return batch
 
         ds = load_dataset("patrickvonplaten/librispeech_asr_dummy", "clean", split="validation")
-        ds = ds.select(range(num_samples)).map(map_to_array)
+        ds = ds.sort("id").select(range(num_samples)).map(map_to_array)
```
Comment (anton-l) [inline tests/test_modeling_speech_to_text.py:727]: _Unrelated to this PR_: the new`datasets` version re-shuffled this dataset, so sorting is needed for reproducibility.

File: tests/test_pipelines_automatic_speech_recognition.py [hunk 5]
```diff
@@ -68,14 +68,14 @@ def test_torch_large(self):
...
-        waveform = np.zeros((34000,))
+        waveform = np.linspace(0, 1, 34000, dtype=np.float32)
         output = speech_recognizer(waveform)
         self.assertEqual(output, {"text": ""})
 
         from datasets import load_dataset
 
-        ds = load_dataset("patrickvonplaten/librispeech_asr_dummy", "clean", split="validation")
-        filename = ds[0]["file"]
+        ds = load_dataset("patrickvonplaten/librispeech_asr_dummy", "clean", split="validation").sort("id")
```
Comment (anton-l) [inline tests/test_pipelines_automatic_speech_recognition.py:77]: _Unrelated to this PR_: the new`datasets` version re-shuffled this dataset, so sorting is needed for reproducibility.

File: tests/test_pipelines_automatic_speech_recognition.py [hunk 6]
```diff
@@ -42,9 +42,9 @@ def test_torch_small(self):
             tokenizer="facebook/s2t-small-mustc-en-fr-st",
             framework="pt",
         )
-        waveform = np.zeros((34000,))
+        waveform = np.linspace(0, 1, 34000, dtype=np.float32)
```
Comment (anton-l) [inline tests/test_pipelines_automatic_speech_recognition.py:45]: Now that the slight variability due to type-casting is eliminated, this needs to be something specific, because the model becomes unstable (non-reproducible in different environments) with all-zero inputs.

File: src/transformers/feature_extraction_sequence_utils.py [hunk 7]
```diff
@@ -163,28 +163,31 @@ def pad(
...
             if is_tf_available() and _is_tensorflow(first_element):
-                return_tensors = "tf" if return_tensors is None else return_tensors
+                return_tensors = "tf"
             elif is_torch_available() and _is_torch(first_element):
-                return_tensors = "pt" if return_tensors is None else return_tensors
-            elif isinstance(first_element, np.ndarray):
-                return_tensors = "np" if return_tensors is None else return_tensors
+                return_tensors = "pt"
+            elif isinstance(first_element, (int, float, list, tuple, np.ndarray)):
+                return_tensors = "np"
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_sequence_utils.py:173]: Nice!

File: src/transformers/file_utils.py [hunk 8]
```diff
@@ -1859,6 +1859,20 @@ def to_py_obj(obj):
         return obj
 
 
+def to_numpy(obj):
+    """
+    Convert a TensorFlow tensor, PyTorch tensor, Numpy array or python list to a Numpy array.
+    """
```
Comment (patrickvonplaten) [inline src/transformers/file_utils.py:1865]: maybe add 

```
    if isinstance(obj, (dict, UserDict)):
        return {k: to_py_obj(v) for k, v in obj.items()}
``` 

as well?

And could you maybe also add a:

```
elif is_flax_available() and _is_jax(obj):
    return np.asarray(obj)
```

Comment (patrickvonplaten) [reply src/transformers/file_utils.py:1865]: we should actually add this to `to_py_obj(obj)` as well

File: src/transformers/feature_extraction_sequence_utils.py [hunk 9]
```diff
@@ -278,42 +280,47 @@ def _pad(
...
+                processed_features[self.model_input_names[0]] = np.pad(
+                    required_input, padding_shape, "constant", constant_values=self.padding_value
+                )
             elif self.padding_side == "left":
                 if return_attention_mask:
-                    processed_features["attention_mask"] = [0] * difference + [1] * len(required_input)
-                processed_features[self.model_input_names[0]] = [
-                    padding_vector for _ in range(difference)
-                ] + required_input
+                    attention_mask = np.zeros(max_length, dtype=np.int64)
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_sequence_utils.py:298]: why `np.int64` here? Think `np.int32` should be good no?

File: src/transformers/feature_extraction_sequence_utils.py [hunk 10]
```diff
@@ -278,42 +280,47 @@ def _pad(
...
         if needs_to_be_padded:
             difference = max_length - len(required_input)
-            padding_vector = self.feature_size * [self.padding_value] if self.feature_size > 1 else self.padding_value
             if self.padding_side == "right":
                 if return_attention_mask:
-                    processed_features["attention_mask"] = [1] * len(required_input) + [0] * difference
-                processed_features[self.model_input_names[0]] = required_input + [
-                    padding_vector for _ in range(difference)
-                ]
+                    attention_mask = np.zeros(max_length, dtype=np.int64)
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_sequence_utils.py:289]: think `np.int32` is good enough here

File: src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/models/speech_to_text/feature_extraction_speech_to_text.py:207]: yes!

File: tests/test_feature_extraction_speech_to_text.py [hunk 11]
```diff
@@ -137,15 +137,9 @@ def test_cepstral_mean_and_variance_normalization(self):
...
-        # TODO(Patrick, Suraj, Anton) - It's surprising that "non-padded/non-numpified" padding
-        # results in quite inaccurate variance computation after (see 5e-1 tolerance)
-        # Issue is filed and PR is underway: https://github.com/huggingface/transformers/issues/13539
-        #        paddings = ["longest", "max_length", "do_not_pad"]
-        #        max_lengths = [None, 16, None]
-        #        var_tolerances = [1e-3, 1e-3, 5e-1]
-        paddings = ["longest", "max_length"]
-        max_lengths = [None, 16]
-        var_tolerances = [1e-3, 1e-3]
+        paddings = ["longest", "max_length", "do_not_pad"]
```
Comment (patrickvonplaten) [inline tests/test_feature_extraction_speech_to_text.py:140]: very nice!

File: tests/test_feature_extraction_speech_to_text.py [hunk 12]
```diff
@@ -137,15 +137,9 @@ def test_cepstral_mean_and_variance_normalization(self):
...
+        max_lengths = [None, 16, None]
+        var_tolerances = [1e-3, 1e-3, 1e-1]
```
Comment (patrickvonplaten) [inline tests/test_feature_extraction_speech_to_text.py:142]: still need `1e-1` tolerance?

Comment (patrickvonplaten) [reply tests/test_feature_extraction_speech_to_text.py:142]: `1e-3` doesn't work?

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me! Left a couple of nits:

- Think we don't have to use `np.int64` -> we are using `tf.int32` everywhere
- Would be nice to add `jax` to the `to_py_obj` and `to_numpy_obj` function
- Do we still need a higher variance for the `no-padding` test for `Speech2Text`?

**Comment by patrickvonplaten** [comment]
Also did you notice a speed-up for larger inputs?

**Review by LysandreJik — APPROVED** [review]
Super cool, LGTM @anton-l!

File: src/transformers/feature_extraction_sequence_utils.py (hunk 9 above)
Comment (anton-l) [reply src/transformers/feature_extraction_sequence_utils.py:298]: Wasn't fully sure about this due to https://github.com/huggingface/transformers/pull/13650#discussion_r712159662

Comment (anton-l) [reply src/transformers/feature_extraction_sequence_utils.py]: You're right though, switched them back to int32

File: tests/test_feature_extraction_speech_to_text.py (hunk 12 above)
Comment (anton-l) [reply tests/test_feature_extraction_speech_to_text.py]: Actually it does work, thanks for catching this! :)

**Comment by anton-l** [comment]
@patrickvonplaten the benchmarking results are pretty promising:

1. Input lengths from 8000 to 16000 (1 sec max), batch size 64, feature_extractor only: 
    - Python: 52.1 ms ± 2.35 ms
    - Numpy: **32.1 ms** ± 1.13 ms

2. Input lengths from 8000 to 160000 (10 sec max), batch size 64, feature_extractor only: 
    - Python: 276 ms ± 950 µs
    - Numpy: **68.2 ms** ± 491 µs

File: src/transformers/file_utils.py (hunk 8 above)
Comment (anton-l) [reply src/transformers/file_utils.py:1867]: Added :+1:

File: tests/test_feature_extraction_speech_to_text.py [hunk 13]
```diff
@@ -137,17 +137,10 @@ def test_cepstral_mean_and_variance_normalization(self):
...
-        # Issue is filed and PR is underway: https://github.com/huggingface/transformers/issues/13539
-        #        paddings = ["longest", "max_length", "do_not_pad"]
-        #        max_lengths = [None, 16, None]
-        #        var_tolerances = [1e-3, 1e-3, 5e-1]
-        paddings = ["longest", "max_length"]
-        max_lengths = [None, 16]
-        var_tolerances = [1e-3, 1e-3]
+        paddings = ["longest", "max_length", "do_not_pad"]
+        max_lengths = [None, 16, None]
+        var_tolerances = [1e-3, 1e-3, 1e-3]
```
Comment (patrickvonplaten) [inline tests/test_feature_extraction_speech_to_text.py:142]:
```suggestion
```

File: tests/test_feature_extraction_speech_to_text.py [hunk 14]
```diff
@@ -137,17 +137,10 @@ def test_cepstral_mean_and_variance_normalization(self):
...
         for max_length, padding, var_tol in zip(max_lengths, paddings, var_tolerances):
```
Comment (patrickvonplaten) [inline tests/test_feature_extraction_speech_to_text.py:143]:
```suggestion
        for max_length, padding in zip(max_lengths, paddings):
```

Comment (patrickvonplaten) [reply tests/test_feature_extraction_speech_to_text.py:143]: think we can delete it completely then :-)

**Comment by patrickvonplaten** [comment]
Great job @anton-l - feel free to merge!
